designs/2024-config-extends/README.md

```diff
@@ -0,0 +1,499 @@
+
+Here's an actual example from [`@eslint/json`](https://github.com/eslint/json):
+
+```js
+const plugin = {
+	meta: {
+		name: "@eslint/json",
+		version: "0.6.0", // x-release-please-version
+	},
+	languages: {
+		json: new JSONLanguage({ mode: "json" }),
+		jsonc: new JSONLanguage({ mode: "jsonc" }),
+		json5: new JSONLanguage({ mode: "json5" }),
+	},
+	rules: {
+		"no-duplicate-keys": noDuplicateKeys,
+		"no-empty-keys": noEmptyKeys,
+	},
+
+    // can't include the plugin here because it must reference `plugin`
+	configs: {},
+};
+
+Object.assign(plugin.configs, {
+	recommended: {
+
+        // now we can reference plugin
+		plugins: { json: plugin },
+		rules: {
+			"json/no-duplicate-keys": "error",
+			"json/no-empty-keys": "error",
+		},
+	},
+});
+
+export default plugin;
+```
+
+Here, we are hardcoding the namespace `json` even though that might not be the namespace that the user assigns to this plugin. This is something we can now address with the use of `extends` because we have the ability to alter the config before inserting it.
+
+Instead of using a hardcoded plugin namespace, plugins can instead use `#` to indicate that they'd like to have the plugin itself included and referenced using the namespace the user assigned. For example, we could rewrite the JSON plugin like this:
+
+```js
+export default {
+	meta: {
+		name: "@eslint/json",
+		version: "0.6.0", // x-release-please-version
+	},
+	languages: {
+		json: new JSONLanguage({ mode: "json" }),
+		jsonc: new JSONLanguage({ mode: "jsonc" }),
+		json5: new JSONLanguage({ mode: "json5" }),
+	},
+	rules: {
+		"no-duplicate-keys": noDuplicateKeys,
+		"no-empty-keys": noEmptyKeys,
+	},
+
+	configs: {
+        recommended: {
+            plugins: { "#": null },
+            rules: {
+                "#/no-duplicate-keys": "error",
+                "#/no-empty-keys": "error",
+            },
+        },
+    },
+};
+```
+
+In an `eslint.config.js` file:
+
+```js
+import json from "@eslint/json";
+
+export default [
+    {
+        plugins: {
+            j: json
+        },
+        files: ["**/*.json"],
+        extends: ["j/recommended"],
+        rules: {
+            "j/no-empty-keys": "warn"
+        }
+    }
+];
+```
+
+Here, the user has assigned the namespace `j` to the JSON plugin. When ESLint loads this config, it sees `extends` with a value of `"j/recommended"`. It looks up the plugin referenced by `j` and then the config named `"recommended"`. It then sees that there's a plugin entry for `"#"` and replaces that with an entry for `j` that points to the JSON plugin. The next step is to look through the config for all the `#` references and replace that with `j` so the references are correct. 
+
+### Implementation Details
```

Comment:
It's worth referencing the implementation that typescript-eslint already has that is used in the ecosystem (first released in February 2024)


https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/typescript-eslint/src/config-helper.ts